### DISCUSSION TITLE
S3 scheme deprecation?

### DISCUSSION THREAD

**Discussion by rschmunk** [Unidata/netcdf-java#1053] [Questions and Answers] [answered]
I have built the netcdfAll jar with the optional S3 module included, and successfully opened some datasets from the AWS S3 archives at registry.opendata.aws. 

However, when I do so, a warning percolates out of the NJ code:
`Use of the s3 scheme is deprecated. Please switch to the cdms3 scheme.`

This seems to suggest that when acquiring an S3 dataset, I should specify the address in the form of `cdms3://bucket/key` rather than `s3://bucket.key`. Right? But if try that, I get an exception:
`java.lang.IllegalStateException: The cdmS3Uri must reference an object - object key is missing.`

Am I mis-understanding that deprecation notice? Or is there something else in play?

**Comment by lesserwhirls** [comment, marked as answer]
Greetings @rschmunk! The way AWS S3 URIs are formatted was too limiting for our needs (other object stores, cloud or not, multiple credential profiles, plus some other netCDF-Java specific things used when scanning buckets), so we came up with the cdms3 URI. If you are moving from _s3://_ to _cdms3://_, the mapping would be:

_s3://bucket/key_

to

_cdms3://aws/bucket?key_

or

_cdms3:bucket?key_

(edited to fix AWS example)

The full details of the _cdms3_ URI can be found at:

https://docs.unidata.ucar.edu/netcdf-java/current/userguide/dataset_urls.html#object-stores

As you will see in the docs, it has more capabilities that the AWS S3 URI.

**Comment by rschmunk** [comment]
Thx, @lesserwhirls . Apparently when I Googled for help about loading S3 via netCDF-Java, I got an [older version](https://docs.unidata.ucar.edu/netcdf-java/5.3/userguide/dataset_urls.html) of that documentation page.

Edit: I succeeded in loading
``cdms3:noaa-goes16?ABI-L1b-RadC/2019/363/21/OR_ABI-L1b-RadC-M6C16_G16_s20193632101189_e20193632103574_c20193632104070.nc``

**Reply by rschmunk** [reply]
Comments in my code now say

```
//  Three examples for the same dataset,
//  s3://noaa-goes16/ABI-L1b-RadC/2019/363/21/OR_ABI-L1b-RadC-M6C16_G16_s20193632101189_e20193632103574_c20193632104070.nc
//  cdms3:noaa-goes16?ABI-L1b-RadC/2019/363/21/OR_ABI-L1b-RadC-M6C16_G16_s20193632101189_e20193632103574_c20193632104070.nc
//  cdms3://aws/noaa-goes16?ABI-L1b-RadC/2019/363/21/OR_ABI-L1b-RadC-M6C16_G16_s20193632101189_e20193632103574_c20193632104070.nc
```

**Reply by lesserwhirls** [reply]
I edited my answer above to fix the cdms3 example...I forgot it's just _cdms3:bucket?key_ for the simplified AWS example (without the URI authority part), or _cdms3://aws/bucket?key_ (with the authority part).

**Reply by rschmunk** [reply]
My latest Panoply release now includes S3 capability.

Now, explaining it to anyone who sees the changelog and asks for help.🙄
